pandas/core/index.py

```diff
@@ -391,6 +395,11 @@ def _shallow_copy(self, values=None, infer=False, **kwargs):
 
         if infer:
             attributes['copy'] = False
+            if self._infer_as_myclass:
```

Comment:
why is ``infer`` not good enough here? (or maybe just reprupose and make ``infer`` a class attribute? (lke what you are calling ``_infer_as_myclass``).  e.g. confusing that we need/have 2